Adding Derived type for parameter dependency

aramirezreyes opened the pull request:
In this branch I will give it a go to implemented the feature described in #358. I have never programmed a macro so I will try to follow #169.  Advice from @sebastianpech is welcome!

aramirezreyes commented:
This is a prototype that I thought of today. To date it does not break any test. 
So far, a ComputedParameter can only depend on 1 pre-existing parameter.

Does this design make sense @Datseris?

Datseris commented:
This does make sense to me! But also, shouldn't we allow the produced parameter to depend in as many parameters as the user wants...?

Just make the field of the parameter a vector of values instead, and if a user gives a single string/symbol make it into a vector anyways.

aramirezreyes commented:
I added this change and some docs. Now the use cases I added to the tests are:

```julia

p = Dict(:α => [1, 2],
    :solver => [SolverA,SolverB],
    :β => ComputedParameter(:α, x -> x^2),
    )


@test Set([ Dict(:α => 1, :solver => SolverA, :β => 1),
    Dict(:α => 2, :solver => SolverA, :β => 4),
    Dict(:α => 1, :solver => SolverB, :β => 1),
    Dict(:α => 2, :solver => SolverB, :β => 4),
    ]) == Set(dict_list(p))

p2 = Dict(:α => [1, 2],
    :β => [10,100],
    :solver => [SolverA,SolverB],
    :γ => ComputedParameter([:α,:β], (x,y) -> x^2 + 2y),
    )

@test Set([ Dict(:α => 1, :solver => SolverA, :β => 10, :γ => 21),
    Dict(:α => 2, :solver => SolverA, :β => 10, :γ => 24),
    Dict(:α => 1, :solver => SolverB, :β => 10, :γ => 21),
    Dict(:α => 2, :solver => SolverB, :β => 10, :γ => 24),
    Dict(:α => 1, :solver => SolverA, :β => 100, :γ => 201),
    Dict(:α => 2, :solver => SolverA, :β => 100, :γ => 204),
    Dict(:α => 1, :solver => SolverB, :β => 100, :γ => 201),
    Dict(:α => 2, :solver => SolverB, :β => 100, :γ => 204),
   ]) == Set(dict_list(p2))
```

Datseris commented:
This is great! You should
0. Consider renaming the whole `ComputedParameter` to just `Derived`. Everything is computed after all :) 
1. Increase minor version in Project.toml
2. Add a changelog entry for this
3. Add a refenrece to `Derived` in the docstring of `dict_list`.
4. Actually add `Derived` to the documentaiton, which means go to the file `run&list.md`, and in the `@docs` block that has `dict_list`, also add `Derived`.

then I'll merge and tag!

Datseris commented on src/dict_list.jl:
```diff
@@ -275,3 +275,82 @@ julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
+    return ComputedParameter([independentP], func)
+end
+
+export ComputedParameter
```
Also please move this to the top of the file with the other export statements

aramirezreyes replied:
Oh yes. Actually, this file did not have the exports at the top. I took the liberty of moving them up.

Datseris commented on src/dict_list.jl:
```diff
@@ -275,3 +275,82 @@ julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
+Receive an array of parameter dictionaries, and for each one, evaluates the computed parameters after the possible combination of parameters has been created.
+"""
+function produce_computed_parameters(dicts)
+    map(dicts) do dict
```
By the way, we shouldn't `map` here. This creates a container that we don't need. We should simple loop over the `dict` in `dicts` and perform the subsequent code!

aramirezreyes replied:
Fixed!

aramirezreyes commented:
I think everything is addressed!

Datseris commented on CHANGELOG.md:
```diff
@@ -1,3 +1,6 @@
+# 2.11.0
+* Add `Derived` type and function to allow `dict_list` to compute some parameters from the value of other parameters.
+* Moved `export` statements to the top of the file in `dict_list.jl`
```
```suggestion

```
This has no reason to be mentioned in the changelog xD

Datseris commented on docs/src/real_world.md:
````diff
@@ -279,6 +279,24 @@ dicts = dict_list(general_args)
 println("Total dictionaries made: ", length(dicts))
 dicts[1]
 ```
+Also, using the type [`ComputedParameter`](@ref), we can have parameters that are computed depending on the value of other parameters:
````
```suggestion
Also, using the type [`Derived`](@ref), we can have parameters that are computed depending on the value of other parameters:
```

Datseris commented on src/dict_list.jl:
```diff
@@ -275,3 +298,85 @@ julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
+    func::Function
+end
+
+"""
+    Derived(independentP <: Union{String,Symbol}, func :: Function)
+Constructs a Derived from a single independent parameter.
+"""
+function Derived(independentP :: Union{String,Symbol}, func :: Function) 
```
weird syntax to have white space between `::`. I/ve never seen it before. Should be without white space.

Datseris commented on src/dict_list.jl:
```diff
@@ -275,3 +298,85 @@ julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
  end
+
+"""
+   Derived(parameters::Vector{Union{String,Symbol}},function::Function)
```
```suggestion
   Derived(parameters::Vector{Union{String,Symbol}}, function::Function)
```

Datseris commented on src/dict_list.jl:
```diff
@@ -275,3 +298,85 @@ julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
+"""
+   Derived(parameters::Vector{Union{String,Symbol}},function::Function)
+Wrap the name(s) of a parameter(s) and a function. After the 
+possible parameter combinations are created, dict_list will replace instances of 
```
```suggestion
possible parameter combinations are created, [`dict_list`](@ref) will replace instances of 
```

Datseris commented on src/dict_list.jl:
```diff
@@ -275,3 +298,85 @@ julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
+
+"""
+   produce_computed_parameter(dicts)
+Receive an array of parameter dictionaries, and for each one, evaluates the computed parameters after the possible combination of parameters has been created.
```
```suggestion
Receive an array of parameter dictionaries, and for each one, evaluate
the computed parameters after the possible combination of 
parameters has been created.
```

Datseris commented on docs/src/real_world.md:
```diff
@@ -279,6 +279,24 @@ dicts = dict_list(general_args)
+    "noise2" => [1.0, ComputedParameter(["noise", "N"], (x,y) -> 2x + y)],
+    "noisy_training" => true,
+    "N" => 100,
+       )
```
```suggestion
)
```

Datseris commented:
thanks, I Think that's a great feature :)

aramirezreyes commented:
Thanks!
